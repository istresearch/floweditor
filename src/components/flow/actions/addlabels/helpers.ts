import { AddLabelsFormState } from 'components/flow/actions/addlabels/AddLabelsForm';
import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { AddLabels, Label } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): AddLabelsFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.add_input_labels) {
    const action = settings.originalAction as AddLabels;
    return {
      labels: {
        value: action.labels.map((label: Label) => {
          if (label.name_match) {
            return { name: label.name_match, expression: true };
          }
          return label;
        })
      },
      valid: true
    };
  }

  return {
    labels: { value: [] },
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  formState: AddLabelsFormState
): AddLabels => {
  const result = {
    type: Types.add_input_labels,
    labels: formState.labels.value.map((label: any) => {
      if (label.expression) {
        return { name_match: label.name };
      }
      return label;
    }),
    uuid: getActionUUID(settings, Types.add_input_labels)
  };
  return result;
};
