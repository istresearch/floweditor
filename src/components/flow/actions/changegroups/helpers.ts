import { Group } from 'flowTypes';
import { Asset, AssetType } from 'store/flowContext';
import { FormState } from 'store/nodeEditor';

export const labelSpecId = 'label';

export interface ChangeGroupsFormState extends FormState {
  groups: any;
  removeAll?: boolean;
}

export const excludeDynamicGroups = (group: any): boolean => {
  return !!group.query;
};

export const mapGroupsToAssets = (groups: Group[]): Asset[] => {
  return groups.map(({ name, uuid, name_match }) => {
    if (name_match) {
      return { name: name_match, id: name_match, type: AssetType.NameMatch };
    }
    return { name, id: uuid, type: AssetType.Group };
  });
};

export const mapAssetsToGroups = (searchResults: Asset[] | null): Group[] => {
  if (!searchResults) {
    return [];
  }

  const groups = searchResults.map(result => {
    if (!result.id || result.id === result.name) {
      return {
        name_match: result.name
      };
    }
    return {
      uuid: result.id,
      name: result.name
    };
  });

  return groups;
};
