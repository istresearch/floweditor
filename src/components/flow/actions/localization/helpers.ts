import { KeyLocalizationFormState } from 'components/flow/actions/localization/KeyLocalizationForm';
import { MsgLocalizationFormState } from 'components/flow/actions/localization/MsgLocalizationForm';
import { Types } from 'config/interfaces';
import { getTypeConfig } from 'config/typeConfigs';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';
import { SendMsg, MsgTemplating, SayMsg } from 'flowTypes';
import { Attachment } from '../sendmsg/attachments';

export const initializeLocalizedKeyForm = (
  settings: NodeEditorSettings
): KeyLocalizationFormState => {
  const keyValues: { [key: string]: StringEntry } = {};
  const localized = settings.localizations[0];
  const action = localized.getObject() as any;

  const keys = settings.originalAction
    ? getTypeConfig(settings.originalAction.type).localizeableKeys || []
    : [];
  keys.forEach((key: string) => {
    keyValues[key] = { value: key in localized.localizedKeys ? action[key] : '' };
  });

  return {
    keyValues,
    valid: true
  };
};

export const initializeLocalizedForm = (settings: NodeEditorSettings): MsgLocalizationFormState => {
  const state: MsgLocalizationFormState = {
    message: { value: '' },
    quickReplies: { value: [] },
    templateVariables: [],
    templating: null,
    audio: { value: null },
    valid: true,
    attachments: []
  };

  // check if our form should use a localized action
  if (
    settings.originalAction &&
    (settings.originalAction.type === Types.send_msg ||
      settings.originalAction.type === Types.say_msg) &&
    settings.localizations &&
    settings.localizations.length > 0
  ) {
    if (settings.originalAction && (settings.originalAction as any).templating) {
      state.templating = (settings.originalAction as any).templating;
      state.templateVariables = state.templating.variables.map((value: string) => {
        return {
          value: ''
        };
      });
    }

    for (const localized of settings.localizations) {
      if (localized.isLocalized()) {
        const localizedObject = localized.getObject() as any;

        if (localizedObject.text) {
          const action = localizedObject as (SendMsg & SayMsg);
          state.message.value = 'text' in localized.localizedKeys ? action.text : '';
          state.audio.value = 'audio_url' in localized.localizedKeys ? action.audio_url : null;
          state.quickReplies.value =
            'quick_replies' in localized.localizedKeys ? action.quick_replies || [] : [];

          const attachments: Attachment[] = [];

          if ('attachments' in localized.localizedKeys) {
            (action.attachments || []).forEach((attachmentString: string) => {
              const splitPoint = attachmentString.indexOf(':');

              const type = attachmentString.substring(0, splitPoint);
              const attachment = {
                type,
                url: attachmentString.substring(splitPoint + 1),
                uploaded: type.indexOf('/') > -1
              };

              attachments.push(attachment);
            });
          }

          state.attachments = attachments;
          state.valid = true;
        }

        if (localizedObject.variables) {
          const templating = localizedObject as MsgTemplating;
          state.templateVariables = templating.variables.map((value: string) => {
            return {
              value: 'variables' in localized.localizedKeys ? value : ''
            };
          });
          state.valid = true;
        }
      }
    }
  }
  return state;
};
