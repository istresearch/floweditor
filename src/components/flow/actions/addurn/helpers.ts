import { AddURNFormState } from 'components/flow/actions/addurn/AddURNForm';
import { getActionUUID } from 'components/flow/actions/helpers';
import { SelectOption } from 'components/form/select/SelectElement';
import { Types } from 'config/interfaces';
import { Scheme, SCHEMES } from 'config/typeConfigs';
import { AddURN } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const getSchemeOptions = (): SelectOption[] => {
  return SCHEMES.map((scheme: Scheme) => {
    return schemeToSelectOption(scheme);
  });
};

export const getSchemeObject = (scheme: string): Scheme => {
  return SCHEMES.find((item: Scheme) => item.scheme === scheme)!;
};

export const getSchemeSelectOption = (scheme: string): SelectOption =>
  schemeToSelectOption(getSchemeObject(scheme));

export const schemeToSelectOption = (scheme: Scheme): SelectOption => {
  return { value: scheme.scheme, name: scheme.path };
};

export const initializeForm = (settings: NodeEditorSettings): AddURNFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.add_contact_urn) {
    const { scheme, path } = settings.originalAction as AddURN;

    return {
      scheme: { value: getSchemeSelectOption(scheme) },
      path: { value: path },
      valid: true
    };
  }

  return {
    scheme: { value: getSchemeSelectOption('tel') },
    path: { value: '' },
    valid: false
  };
};

export const stateToAction = (settings: NodeEditorSettings, formState: AddURNFormState): AddURN => {
  return {
    type: Types.add_contact_urn,
    uuid: getActionUUID(settings, Types.add_input_labels),
    scheme: formState.scheme.value.value,
    path: formState.path.value
  };
};
