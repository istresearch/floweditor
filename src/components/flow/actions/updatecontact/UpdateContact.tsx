import { getLanguageForCode } from 'components/flow/actions/updatecontact/helpers';
import i18n from 'config/i18n';
import { Types } from 'config/interfaces';
import {
  SetContactAttribute,
  SetContactChannel,
  SetContactLanguage,
  SetContactName,
  SetContactStatus
} from 'flowTypes';
import * as React from 'react';
import { emphasize } from 'utils';
const styles = require('components/shared.module.scss');

const withEmph = (text: string, emph: boolean) => (emph ? emphasize(text) : text);

export const renderSetText = (
  name: string,
  value: string,
  emphasizeName: boolean = false
): JSX.Element => {
  if (value) {
    return (
      <div className={`${styles.node_asset}`}>
        Set {withEmph(name, emphasizeName)} to {emphasize(value)}.
      </div>
    );
  } else {
    return (
      <div>
        {i18n.t('forms.clear', 'Clear')} {withEmph(name, emphasizeName)}.
      </div>
    );
  }
};

const UpdateContactComp: React.SFC<SetContactAttribute> = (
  action: SetContactAttribute
): JSX.Element => {
  if (action.type === Types.set_contact_field) {
    return renderSetText(action.field.name, action.value, true);
  }

  if (action.type === Types.set_contact_channel) {
    const setContactAction = action as SetContactChannel;
    return renderSetText(
      'channel',
      setContactAction.channel ? setContactAction.channel.name : null,
      false
    );
  }

  if (action.type === Types.set_contact_language) {
    const setLanguageAction = action as SetContactLanguage;
    return renderSetText(
      'language',
      getLanguageForCode(setLanguageAction.language, (action as any).languages),
      false
    );
  }

  if (action.type === Types.set_contact_status) {
    return renderSetText('status', (action as SetContactStatus).status, false);
  }

  if (action.type === Types.set_contact_name) {
    return renderSetText('name', (action as SetContactName).name, false);
  }

  return null;
};

export default UpdateContactComp;
