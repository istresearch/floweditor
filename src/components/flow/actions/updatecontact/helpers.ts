import { getActionUUID } from 'components/flow/actions/helpers';
import {
  CHANNEL_PROPERTY,
  LANGUAGE_PROPERTY,
  NAME_PROPERTY,
  STATUS_PROPERTY
} from 'components/flow/props';
import { Types, ContactStatus } from 'config/interfaces';
import { getTypeConfig } from 'config/typeConfigs';
import {
  Field,
  SetContactAttribute,
  SetContactChannel,
  SetContactField,
  SetContactLanguage,
  SetContactName,
  SetContactStatus
} from 'flowTypes';
import { Asset, AssetMap, AssetStore, AssetType, REMOVE_VALUE_ASSET } from 'store/flowContext';
import {
  FormState,
  NodeEditorSettings,
  StringEntry,
  SelectOptionEntry,
  FormEntry
} from 'store/nodeEditor';
import {
  CONTACT_STATUS_OPTIONS,
  CONTACT_STATUS_ACTIVE
} from 'components/flow/actions/updatecontact/UpdateContactForm';

export interface UpdateContactFormState extends FormState {
  type: Types;
  name: StringEntry;
  channel: FormEntry;
  language: FormEntry;
  status: SelectOptionEntry;
  field: FormEntry;
  fieldValue: StringEntry;
}

export const initializeForm = (
  settings: NodeEditorSettings,
  assetStore: AssetStore
): UpdateContactFormState => {
  const state: UpdateContactFormState = {
    type: Types.set_contact_name,
    valid: false,
    name: { value: '' },
    channel: { value: null },
    language: { value: null },
    status: { value: CONTACT_STATUS_ACTIVE },
    field: { value: NAME_PROPERTY },
    fieldValue: { value: '' }
  };

  if (settings.originalAction) {
    const originalType = settings.originalAction.type;
    // these have aliases, so compare the config we resolve to
    if (getTypeConfig(originalType) === getTypeConfig(Types.set_contact_field)) {
      state.type = originalType;

      switch (originalType) {
        case Types.set_contact_field:
          const fieldAction = settings.originalAction as SetContactField;
          state.field = { value: { key: fieldAction.field.key, label: fieldAction.field.name } };
          state.fieldValue = { value: fieldAction.value };
          state.valid = true;
          return state;
        case Types.set_contact_channel:
          const channelAction = settings.originalAction as SetContactChannel;
          state.field = { value: CHANNEL_PROPERTY };
          state.channel = {
            value: channelAction.channel ? channelAction.channel : REMOVE_VALUE_ASSET
          };
          state.valid = true;
          return state;
        case Types.set_contact_language:
          const languageAction = settings.originalAction as SetContactLanguage;
          state.field = { value: LANGUAGE_PROPERTY };
          state.valid = true;
          state.language = {
            value: languageAction.language
              ? {
                  iso: languageAction.language,
                  name: getLanguageForCode(languageAction.language, assetStore.languages.items)
                }
              : REMOVE_VALUE_ASSET
          };
          return state;
        case Types.set_contact_status:
          const statusAction = settings.originalAction as SetContactStatus;
          state.field = { value: STATUS_PROPERTY };
          state.valid = true;
          state.status = {
            value: CONTACT_STATUS_OPTIONS.find(o => o.value === statusAction.status)
          };
          return state;
        case Types.set_contact_name:
          const nameAction = settings.originalAction as SetContactName;
          state.valid = true;
          state.name = {
            value: nameAction.name
          };
          return state;
      }
    }
  }

  // default is updating name
  return state;
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: UpdateContactFormState
): SetContactAttribute => {
  /* istanbul ignore else */
  const field = state.field.value;
  if (state.type === Types.set_contact_field) {
    return {
      uuid: getActionUUID(settings, Types.set_contact_field),
      type: state.type,
      field: { name: field.label, key: field.key },
      value: state.fieldValue.value
    };
  } else if (state.type === Types.set_contact_channel) {
    if (state.channel.value.type === REMOVE_VALUE_ASSET.type) {
      return {
        uuid: getActionUUID(settings, Types.set_contact_channel),
        type: state.type
      } as any;
    }
    return {
      uuid: getActionUUID(settings, Types.set_contact_channel),
      type: state.type,
      channel: {
        uuid: state.channel.value.uuid,
        name: state.channel.value.name
      }
    };
  } else if (state.type === Types.set_contact_language) {
    if (state.language.value.type === REMOVE_VALUE_ASSET.type) {
      return {
        uuid: getActionUUID(settings, Types.set_contact_language),
        type: state.type
      } as any;
    }
    return {
      uuid: getActionUUID(settings, Types.set_contact_language),
      type: state.type,
      language: state.language.value.iso
    };
  } else if (state.type === Types.set_contact_status) {
    return {
      uuid: getActionUUID(settings, Types.set_contact_status),
      type: state.type,
      status: state.status.value.value as ContactStatus
    };
  } else if (state.type === Types.set_contact_name) {
    return {
      uuid: getActionUUID(settings, Types.set_contact_name),
      type: state.type,
      name: state.name.value
    };
  }
};

export const sortFieldsAndProperties = (a: any, b: any): number => {
  const aType = a.type || '';
  const bType = b.type || '';

  const aName = a.name || a.label || '';
  const bName = b.name || b.label || '';

  // Name always goes first
  /* istanbul ignore else */
  if (a.id === NAME_PROPERTY.id && b.id !== NAME_PROPERTY.id) {
    return -1;
  } else if (b.id === NAME_PROPERTY.id && a.id !== NAME_PROPERTY.id) {
    return 1;
  }

  if (aType === bType) {
    return aName.localeCompare(bName);
  }

  if (aType === AssetType.Scheme) {
    return 1;
  }

  if (bType === AssetType.Scheme) {
    return -1;
  }

  // go with alpha-sort for everthing else
  else if (aType !== bType) {
    if (aType === AssetType.ContactProperty) {
      return -1;
    }

    if (bType === AssetType.ContactProperty) {
      return 1;
    }
  }
  // non-name non-fields go last
  return aName.localeCompare(bName);
};

export const fieldToAsset = (field: Field = { key: '', name: '' }): Asset => ({
  id: field.key,
  name: field.name,
  type: AssetType.Field
});

export const assetToField = (asset: Asset): Field => ({
  key: asset.id,
  name: asset.name
});

export const assetToChannel = (asset: Asset): any => {
  if (asset.id === REMOVE_VALUE_ASSET.id) {
    return {};
  }

  return {
    uuid: asset.id,
    name: asset.name
  };
};

export const getLanguageForCode = (code: string, languages: AssetMap) => {
  let lang = code;
  if (languages && lang in languages) {
    lang = languages[lang].name;
  }
  return lang;
};

export const getName = (asset: any): string => {
  return asset.label || asset.name;
};
