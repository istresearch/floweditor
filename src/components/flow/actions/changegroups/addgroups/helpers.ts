import { ChangeGroupsFormState } from 'components/flow/actions/changegroups/helpers';
import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { ChangeGroups, Group } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): ChangeGroupsFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.add_contact_groups) {
    const action = settings.originalAction as ChangeGroups;
    return {
      groups: {
        value: action.groups.map((group: Group) => {
          if (group.name_match) {
            return { name: group.name_match, expression: true };
          }
          return group;
        })
      },
      valid: true
    };
  }

  return {
    groups: { value: null },
    valid: false
  };
};

export const stateToAction = (
  nodeSettings: NodeEditorSettings,
  state: ChangeGroupsFormState
): ChangeGroups => {
  return {
    type: Types.add_contact_groups,
    groups: state.groups.value.map((group: any) => {
      if (group.expression) {
        return { name_match: group.name };
      }
      return group;
    }),
    uuid: getActionUUID(nodeSettings, Types.add_contact_groups)
  };
};
