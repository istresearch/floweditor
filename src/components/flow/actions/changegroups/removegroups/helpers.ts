import { ChangeGroupsFormState } from 'components/flow/actions/changegroups/helpers';
import { getActionUUID } from 'components/flow/actions/helpers';
import { Types } from 'config/interfaces';
import { RemoveFromGroups } from 'flowTypes';
import { NodeEditorSettings } from 'store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): ChangeGroupsFormState => {
  if (settings.originalAction && settings.originalAction.type === Types.remove_contact_groups) {
    const action = settings.originalAction as RemoveFromGroups;

    const groups = action.groups || [];
    return {
      groups: { value: groups },
      removeAll: groups.length === 0 || action.all_groups,
      valid: true
    };
  }

  return {
    groups: { value: null },
    removeAll: false,
    valid: false
  };
};

export const stateToAction = (
  settings: NodeEditorSettings,
  state: ChangeGroupsFormState
): RemoveFromGroups => {
  return {
    type: Types.remove_contact_groups,
    groups: state.removeAll
      ? []
      : state.groups.value.map((group: any) => {
          return { uuid: group.uuid, name: group.name };
        }),
    all_groups: !!state.removeAll,
    uuid: getActionUUID(settings, Types.remove_contact_groups)
  };
};
