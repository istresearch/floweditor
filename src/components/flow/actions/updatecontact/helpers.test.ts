import { sortFieldsAndProperties } from 'components/flow/actions/updatecontact/helpers';
import { getContactProperties } from 'components/helpers';

describe('UpdateContact.helpers', () => {
  it('should sort options', () => {
    const contactProperties = getContactProperties();
    expect(contactProperties.sort(sortFieldsAndProperties)).toMatchSnapshot();
  });
});
