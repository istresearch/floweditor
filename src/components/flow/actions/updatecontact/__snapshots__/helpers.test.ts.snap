// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateContact.helpers should sort options 1`] = `
Array [
  Object {
    "id": "name",
    "name": "Name",
    "type": "property",
  },
  Object {
    "id": "channel",
    "name": "Channel",
    "type": "property",
  },
  Object {
    "id": "language",
    "name": "Language",
    "type": "property",
  },
  Object {
    "id": "status",
    "name": "Status",
    "type": "property",
  },
]
`;
